File: python-sanic/calc/calc.py
import cexprtk
from sanic import Sanic
from sanic.response import text

WELCOME: str = "Welcome to calc service\nHTTP POST your expression\n"

SYMBOL_TABLE = cexprtk.Symbol_Table(
    {
        "pi": 3.14159265358979323846264338327950288419716939937510,
        "e": 2.71828182845904523536028747135266249775724709369996,
    }
)

NAN = "NaN"


def create_calc(app_name: str, is_testing: bool) -> Sanic:
    calc = Sanic(app_name)
    if is_testing:
        calc.ctx.test_mode = True

    @calc.get("/<path:path>")
    async def welcome(request, path):
        return text(WELCOME)

    @calc.post("/<path:path>", ignore_body=False)
    async def evaluate(request, path):
        result = NAN
        try:
            if not request.body:
                await request.receive_body()

            body: str = request.body.decode("utf-8")
            expression = cexprtk.Expression(body, SYMBOL_TABLE)
            result = str(expression())
        except:
            pass
        return text(result)

    return calc


calc = create_calc("__main__", False)

if __name__ == "__main__":
    calc.run(port=8080)

File: deno-oak/calc.ts
import { Application, Context } from "@oak/oak";
import { evaluate, ResultSet } from "mathjs";

Deno.addSignalListener("SIGTERM", () => {
  console.log("SIGTERM signal received.");
  Deno.exit(0);
});

Deno.addSignalListener("SIGINT", () => {
  console.log("SIGINT signal received.");
  Deno.exit(0);
});

const WELCOME: string =
  "Welcome to calc service\nHTTP POST your expression / (via mathjs)";
const NAN_STR: string = "NaN";
const EXPRTK: string = "exprtk";
const C_EXPRTK_ADAPTER_NAME: string = "libc-exprtk-adapter.so";
const TEXT_ENCODER: TextEncoder = new TextEncoder();

const C_EXPRTK_ADAPTER = Deno.dlopen(C_EXPRTK_ADAPTER_NAME, {
  calculate: { parameters: ["buffer"], result: "f64" },
});

const HTTP_PORT: number = Deno.env.has("HTTP_PORT")
  ? parseInt(Deno.env.get("HTTP_PORT")!)
  : 8080;

const viaMathJs = (expression: string): number => {
  try {
    let result = evaluate(expression);
    if (result != undefined && result != null) {
      if (result.entries) {
        const entries = (result as ResultSet).entries;
        if (entries.length > 0 && entries[0] instanceof Number) {
          return entries[0] as number;
        }
      } else {
        return result as number;
      }
    }
  } catch {}
  return Number.NaN;
};

const viaExprtk = (expr: string): number => {
  try {
    const c_string_buf = TEXT_ENCODER.encode(expr + "\0");
    return C_EXPRTK_ADAPTER.symbols.calculate(c_string_buf);
  } catch {}
  return Number.NaN;
};

const handler = async (
  ctx: Context<Record<string, any>, Record<string, any>>,
): Promise<void> => {
  if (ctx.request.method === "POST") {
    const expr: string = await ctx.request.body.text();
    let result: string = NAN_STR;
    const url: string = ctx.request.url.toString();
    if (url.includes(EXPRTK)) {
      result = "" + viaExprtk(expr);
    } else {
      result = "" + viaMathJs(expr);
    }
    ctx.response.body = result;
  } else {
    ctx.response.body = WELCOME;
  }
};

const app = new Application();
app.use(handler);
await app.listen({ hostname: "0.0.0.0", port: HTTP_PORT });
